### PR TITLE
Add gateway APM support

### PR DESCRIPTION
**Pull request by ninpan-ms**
---
Add new arguments `--apm-types`, `--properties` and `--secrets` for `az spring gateway update`

### General Guidelines

- [X] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [X] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repository and upgrade the version in the pull request but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
spring

File: src/spring/azext_spring/_params.py [hunk 1]
```diff
@@ -718,6 +718,13 @@ def prepare_logs_argument(c):
         c.argument('api_doc_location', arg_group='API metadata', help="Location of additional documentation for the APIs available on the Gateway instance.")
         c.argument('api_version', arg_group='API metadata', help="Version of APIs available on this Gateway instance.")
         c.argument('server_url', arg_group='API metadata', help="Base URL that API consumers will use to access APIs on the Gateway instance.")
+        c.argument('apm_types', nargs='*',
+                   help="Space-separated list of APM integrated with Gateway, Allowed values: [\"ApplicationInsights\", \"AppDynamics\", \"Dynatrace\", \"NewRelic\", \"ElasticAPM\"].")
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_params.py:722]: Could we define the allowed values as enumeration values? In addition, can we get these values from the Python SDK?

File: src/spring/azext_spring/_validators_enterprise.py [hunk 2]
```diff
@@ -215,6 +218,33 @@ def _validate_sso(namespace):
         namespace.scope = namespace.scope.split(",") if namespace.scope else []
 
 
+def _validate_gateway_apm_types(namespace):
+    if namespace.apm_types is None:
+        return
+    allowedApmTypes = ["ApplicationInsights", "AppDynamics", "Dynatrace", "NewRelic", "ElasticAPM"]
+    for type in namespace.apm_types:
+        if (type not in allowedApmTypes):
+            raise InvalidArgumentValueError("Allowed APM types are \"ApplicationInsights\", \"AppDynamics\", \"Dynatrace\", \"NewRelic\", \"ElasticAPM\".")
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_validators_enterprise.py:227]: Same as above

File: src/spring/azext_spring/_params.py [hunk 3]
```diff
@@ -718,6 +718,13 @@ def prepare_logs_argument(c):
         c.argument('api_doc_location', arg_group='API metadata', help="Location of additional documentation for the APIs available on the Gateway instance.")
         c.argument('api_version', arg_group='API metadata', help="Version of APIs available on this Gateway instance.")
         c.argument('server_url', arg_group='API metadata', help="Base URL that API consumers will use to access APIs on the Gateway instance.")
+        c.argument('apm_types', nargs='*',
+                   help="Space-separated list of APM integrated with Gateway.")
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_params.py:722]:
```suggestion
        c.argument('apm_types', nargs='*',
                   help="Space-separated list of APM integrated with Gateway.", arg_type=get_enum_type(allowedApmTypes))
```

Comment (zhoxing-ms) [reply src/spring/azext_spring/_params.py:722]: I believe the `allowedApmTypes` enumeration can be reused here

File: src/spring/azext_spring/_validators_enterprise.py [hunk 4]
```diff
@@ -215,6 +220,37 @@ def _validate_sso(namespace):
...
+    if namespace.apm_types is None:
+        return
+    allowedApmTypes = [v20221101_preview_AppPlatformEnums.GatewayApmType.APPLICATION_INSIGHTS,
+                       v20221101_preview_AppPlatformEnums.GatewayApmType.APP_DYNAMICS,
+                       v20221101_preview_AppPlatformEnums.GatewayApmType.DYNATRACE,
+                       v20221101_preview_AppPlatformEnums.GatewayApmType.NEW_RELIC,
+                       v20221101_preview_AppPlatformEnums.GatewayApmType.ELASTIC_APM]
+    for type in namespace.apm_types:
+        if (type not in allowedApmTypes):
+            raise InvalidArgumentValueError("Allowed APM types are " + ','.join(allowedApmTypes))
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_validators_enterprise.py:233]: How about using the `v20221101_preview_AppPlatformEnums.GatewayApmType` enumeration from Python SDK directly?

File: src/spring/azext_spring/vendored_sdks/appplatform/v2022_11_01_preview/models/_app_platform_management_client_enums.py [hunk 5]
```diff
@@ -437,3 +437,13 @@ class Type(str, Enum, metaclass=CaseInsensitiveEnumMeta):
...
+
+
+class GatewayApmType(str, Enum, metaclass=CaseInsensitiveEnumMeta):
+    """Spring Cloud Gateway APM Type."""
+
+    APPLICATION_INSIGHTS = "ApplicationInsights"
+    APP_DYNAMICS = "AppDynamics"
+    DYNATRACE = "Dynatrace"
+    NEW_RELIC = "NewRelic"
+    ELASTIC_APM = "ElasticAPM"
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/vendored_sdks/appplatform/v2022_11_01_preview/models/_app_platform_management_client_enums.py:449]: May I ask whether this enumeration is automatically generated by the Python SDK or manually added by you?

File: src/spring/azext_spring/_params.py [hunk 6]
```diff
@@ -41,6 +41,8 @@
     import _app_platform_management_client_enums as v20220101_preview_AppPlatformEnums
 from .vendored_sdks.appplatform.v2022_01_01_preview.models._app_platform_management_client_enums import SupportedRuntimeValue, TestKeyType
 from .vendored_sdks.appplatform.v2022_09_01_preview.models._app_platform_management_client_enums import BackendProtocol, SessionAffinity
+from .vendored_sdks.appplatform.v2022_11_01_preview.models \
+    import _app_platform_management_client_enums as v20221101_preview_AppPlatformEnums
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_params.py:45]:
```suggestion
    import _app_platform_management_client_enums as AppPlatformEnums
```

File: src/spring/azext_spring/_params.py [hunk 7]
```diff
@@ -718,6 +720,13 @@ def prepare_logs_argument(c):
         c.argument('api_doc_location', arg_group='API metadata', help="Location of additional documentation for the APIs available on the Gateway instance.")
         c.argument('api_version', arg_group='API metadata', help="Version of APIs available on this Gateway instance.")
         c.argument('server_url', arg_group='API metadata', help="Base URL that API consumers will use to access APIs on the Gateway instance.")
+        c.argument('apm_types', nargs='*',
+                   help="Space-separated list of APM integrated with Gateway. Allowed values are: " + ', '.join(list(v20221101_preview_AppPlatformEnums.ApmType)))
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_params.py:724]:
```suggestion
                   help="Space-separated list of APM integrated with Gateway. Allowed values are: " + ', '.join(list(AppPlatformEnums.ApmType)))
```

File: src/spring/azext_spring/_validators_enterprise.py [hunk 8]
```diff
@@ -13,6 +13,8 @@
                                        MutuallyExclusiveArgumentError)
 from azure.core.exceptions import ResourceNotFoundError
 from knack.log import get_logger
+from .vendored_sdks.appplatform.v2022_11_01_preview.models \
+    import _app_platform_management_client_enums as v20221101_preview_AppPlatformEnums
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_validators_enterprise.py:17]:
```suggestion
    import _app_platform_management_client_enums as AppPlatformEnums
```

File: src/spring/azext_spring/_validators_enterprise.py [hunk 9]
```diff
@@ -215,6 +220,32 @@ def _validate_sso(namespace):
         namespace.scope = namespace.scope.split(",") if namespace.scope else []
 
 
+def _validate_gateway_apm_types(namespace):
+    if namespace.apm_types is None:
+        return
+    for type in namespace.apm_types:
+        if (type not in list(v20221101_preview_AppPlatformEnums.ApmType)):
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_validators_enterprise.py:227]: Same as above

File: src/spring/azext_spring/_validators_enterprise.py [hunk 10]
```diff
@@ -215,6 +220,32 @@ def _validate_sso(namespace):
...
+            raise InvalidArgumentValueError("Allowed APM types are: " + ', '.join(list(v20221101_preview_AppPlatformEnums.ApmType)))
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/_validators_enterprise.py:228]:
```suggestion
        if (type not in list(AppPlatformEnums.ApmType)):
            raise InvalidArgumentValueError("Allowed APM types are: " + ', '.join(list(AppPlatformEnums.ApmType)))
```

File: src/spring/azext_spring/_params.py (hunk 6 above)
Comment (zhoxing-ms) [inline src/spring/azext_spring/_params.py:45]: Could you make the way of importing enumeration consistent?

File: src/spring/azext_spring/_validators_enterprise.py (hunk 10 above)
Comment (zhoxing-ms) [inline src/spring/azext_spring/_validators_enterprise.py:228]: Same as above

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**